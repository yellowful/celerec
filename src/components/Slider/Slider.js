import React, { useState } from 'react';
import Img1 from './url1.png'
import Img2 from './url2.png'
import Img3 from './url3.png'
import Img4 from './file1.png'
import Img5 from './file2.png'
import Img6 from './file3.png'
import Img7 from './file4.png'
import Img8 from './file5.png'

// 給FormSubmit用的
// 顯示使用的動畫
const Slider = () => {
    // 使用8張圖片，以產生輪播效果
    const imageArray = [Img1, Img2, Img3, Img4, Img5, Img6, Img7, Img8]
    // 記錄滑鼠是否滑入輪播區的狀態
    const [slideState, setSlideState] = useState('')
    // 滑鼠如果滑進輪播區，輪播暫停
    const onMouseEnter = () => { setSlideState('stop-slide') }
    // 滑鼠如果滑出輪播區，輪播繼續
    const onMouseLeave = () => { setSlideState('') }
    // 手機touch輪播區，就改變輪播狀態
    const onTouchEnd = () => {
        if (slideState) {
            setSlideState('')
        } else {
            setSlideState('stop-slide')
        }
    }
    // 記錄截圖是否已被載入的狀態
    // 初始值是一個8個item的array，每一個item都是false
    const [isLoaded,setLoadedState] = useState(new Array(8).fill(false))
    // 如果圖片被載入了，就把對應的載入狀態改成true
    // 是一個closure，會記得各別i是多少，各別的newArray是多少，等於有各自的onLoad
    const onLoad = (i) => () => {
        // 將狀態複制一份，以免動到原狀態
        const newArray = isLoaded.slice();
        // 新狀態對應的item改成true
        newArray[i] = true;
        // 設定狀態存檔
        setLoadedState(newArray);
    }

    return (
        <div className="w-100">
            {/* 
                coverflow設定children都重疊在一起
                並設定延遲的秒數
                slideState如果是stop-slide的話就會暫停
             */}
            <div
                id="cover"
                className={`coverflow ${slideState}`}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
                onTouchEnd={onTouchEnd}
            >
                {
                    // render出8張截圖
                    // 如果isLoaded的array裡面，有false的話，就是有圖片還沒載入完成，這時候就把className設成dn，也就是把圖片先隱藏起來不要顯示
                    // onLoad是一個closure,會把對應的載入狀態改成true
                    imageArray.map((image, i) => {
                        return (
                            <div key={i}>
                                <img 
                                    className={
                                        isLoaded.includes(false) ?
                                            'dn'
                                            :
                                            ''
                                    }
                                    src={image} 
                                    alt={`slide ${i}`} 
                                    onLoad={onLoad(i)}
                                />
                            </div>
                        )
                    })
                }
            </div>
        </div>
    )
}

export default Slider
