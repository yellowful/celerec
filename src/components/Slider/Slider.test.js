import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react'
import Slider from './Slider';

// 測試滑鼠移動到相片上，相片會停止輪播
describe('hover to stop slider', () => {
  // 預期符合快照
  it('should match snapshot', () => {
    const { asFragment } = render(<Slider />);
    expect(asFragment()).toMatchSnapshot();
  })
  // 預期滑鼠移動到相片上，相片會停止輪播
  it('should have class stop', () => {
    render(<Slider />);
    // 滑鼠還沒移到相片上的，相片不會停止輪播
    expect(screen.queryByTitle(/slide/i)).not.toHaveClass('stop-slide');
    // 滑鼠移到相片上
    fireEvent.mouseEnter(screen.queryAllByAltText(/slide/i)[0]);
    // 相片停止輪播
    expect(screen.getByTitle(/slide/i)).toHaveClass('stop-slide');
    // 滑鼠移出相片
    fireEvent.mouseOut(screen.queryAllByAltText(/slide/i)[0]);
    // 相片繼續輪播
    expect(screen.queryByTitle(/slide/i)).not.toHaveClass('stop-slide');
  })

})