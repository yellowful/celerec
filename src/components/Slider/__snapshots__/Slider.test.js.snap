// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hover to stop slider should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-100"
  >
    <div
      class="coverflow "
      id="cover"
      title="slide"
    >
      <div>
        <img
          alt="slide 0"
          class="dn"
          src="url1.png"
        />
      </div>
      <div>
        <img
          alt="slide 1"
          class="dn"
          src="url2.png"
        />
      </div>
      <div>
        <img
          alt="slide 2"
          class="dn"
          src="url3.png"
        />
      </div>
      <div>
        <img
          alt="slide 3"
          class="dn"
          src="file1.png"
        />
      </div>
      <div>
        <img
          alt="slide 4"
          class="dn"
          src="file2.png"
        />
      </div>
      <div>
        <img
          alt="slide 5"
          class="dn"
          src="file3.png"
        />
      </div>
      <div>
        <img
          alt="slide 6"
          class="dn"
          src="file4.png"
        />
      </div>
      <div>
        <img
          alt="slide 7"
          class="dn"
          src="file5.png"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
